pkg/utils/deployment_controller.go

```diff
@@ -0,0 +1,114 @@
+	"github.com/openshift/library-go/pkg/operator/loglevel"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type DeploymentOptions struct {
+	Required       *appsv1.Deployment
+	ControllerName string
+	OpStatus       *operatorapi.OperatorStatus
+	EventRecorder  events.Recorder
+	KubeClient     kubernetes.Interface
+	OperatorClient v1helpers.OperatorClient
+	TargetVersion  string
+	VersionGetter  status.VersionGetter
+	VersionName    string
+}
+
+func CreateDeployment(depOpts DeploymentOptions) (*appsv1.Deployment, error) {
+	lastGeneration := resourcemerge.ExpectedDeploymentGeneration(depOpts.Required, depOpts.OpStatus.Generations)
+	deployment, _, err := resourceapply.ApplyDeployment(depOpts.KubeClient.AppsV1(), depOpts.EventRecorder, depOpts.Required, lastGeneration)
+	if err != nil {
+		// This will set Degraded condition
+		return nil, err
+	}
+
+	// Available: at least one replica is running
+	deploymentAvailable := operatorapi.OperatorCondition{
+		Type: depOpts.ControllerName + operatorapi.OperatorStatusTypeAvailable,
+	}
+	if deployment.Status.AvailableReplicas > 0 {
+		deploymentAvailable.Status = operatorapi.ConditionTrue
+	} else {
+		deploymentAvailable.Status = operatorapi.ConditionFalse
+		deploymentAvailable.Reason = "Deploying"
+		deploymentAvailable.Message = "Waiting for a Deployment pod to start"
+	}
+
+	// Not progressing: all replicas are at the latest version && Deployment generation matches
+	deploymentProgressing := operatorapi.OperatorCondition{
+		Type: depOpts.ControllerName + operatorapi.OperatorStatusTypeProgressing,
+	}
+	if deployment.Status.ObservedGeneration != deployment.Generation {
+		deploymentProgressing.Status = operatorapi.ConditionTrue
+		deploymentProgressing.Reason = "NewGeneration"
+		msg := fmt.Sprintf("desired generation %d, current generation %d", deployment.Generation, deployment.Status.ObservedGeneration)
+		deploymentProgressing.Message = msg
+	} else {
+		if deployment.Spec.Replicas != nil {
+			if deployment.Status.UpdatedReplicas == *deployment.Spec.Replicas {
+				deploymentProgressing.Status = operatorapi.ConditionFalse
+				// All replicas were updated, set the version
+				depOpts.VersionGetter.SetVersion(depOpts.VersionName, depOpts.TargetVersion)
+			} else {
+				msg := fmt.Sprintf("%d out of %d pods running", deployment.Status.UpdatedReplicas, *deployment.Spec.Replicas)
+				deploymentProgressing.Status = operatorapi.ConditionTrue
+				deploymentProgressing.Reason = "Deploying"
+				deploymentProgressing.Message = msg
+			}
+		}
+	}
+
+	depOpts.OpStatus.ReadyReplicas = deployment.Status.ReadyReplicas
```

Comment:
CSI driver will replace nr. of replica of problem-detector deployment and a vice versa. IMO, one of the deployments (or even both) should ignore `ReadyReplicas` completely. The field is not very usable anyway.

Reply:
I removed setting that field.